app/code/Magento/InventorySales/Model/ReturnProcessor/GetSourceDeductionRequestFromSourceSelection.php

```diff
@@ -0,0 +1,133 @@
+     * @param SourceDeductionRequestInterfaceFactory $sourceDeductionRequestFactory
+     * @param SalesChannelInterfaceFactory $salesChannelFactory
+     * @param SalesEventInterfaceFactory $salesEventFactory
+     * @param WebsiteRepositoryInterface $websiteRepository
+     */
+    public function __construct(
+        ItemToDeductInterfaceFactory $itemToDeductFactory,
+        SourceDeductionRequestInterfaceFactory $sourceDeductionRequestFactory,
+        SalesChannelInterfaceFactory $salesChannelFactory,
+        SalesEventInterfaceFactory $salesEventFactory,
+        WebsiteRepositoryInterface $websiteRepository
+    ) {
+        $this->itemToDeductFactory = $itemToDeductFactory;
+        $this->sourceDeductionRequestFactory = $sourceDeductionRequestFactory;
+        $this->salesChannelFactory = $salesChannelFactory;
+        $this->salesEventFactory = $salesEventFactory;
+        $this->websiteRepository = $websiteRepository;
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @param SourceSelectionResultInterface $sourceSelectionResult
+     * @return array|SourceDeductionRequestInterface[]
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function execute(
+        OrderInterface $order,
+        SourceSelectionResultInterface $sourceSelectionResult
+    ): array {
+        $websiteId = (int)$order->getStore()->getWebsiteId();
+
+        $sourceDeductionRequests = [];
+        $websiteCode = $this->websiteRepository->getById($websiteId)->getCode();
+        $salesChannel = $this->salesChannelFactory->create([
+            'data' => [
+                'type' => SalesChannelInterface::TYPE_WEBSITE,
+                'code' => $websiteCode
+            ]
+        ]);
+
+        $salesEvent = $this->salesEventFactory->create([
+            'type' => SalesEventInterface::EVENT_CREDITMEMO_CREATED,
+            'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
+            'objectId' => (string)$order->getEntityId()
+        ]);
+
+        foreach ($this->getItemsPerSource($sourceSelectionResult->getSourceSelectionItems()) as $sourceCode => $items) {
+            /** @var SourceDeductionRequestInterface[] $sourceDeductionRequests */
+            $sourceDeductionRequests[] = $this->sourceDeductionRequestFactory->create([
+                'sourceCode' => $sourceCode,
+                'items' => $items,
+                'salesChannel' => $salesChannel,
+                'salesEvent' => $salesEvent
+            ]);
+        }
+
+        return $sourceDeductionRequests;
+    }
+
+    /**
+     * @param SourceSelectionItemInterface[] $sourceSelectionItems
+     * @return ItemToDeductInterface[]
+     */
+    private function getItemsPerSource(array $sourceSelectionItems): array
+    {
+        $itemsPerSource = [];
+        foreach ($sourceSelectionItems as $sourceSelectionItem) {
+            if ($sourceSelectionItem->getQtyToDeduct() < 0.000001) {
```

Comment:
I'd suggest using `bccomp` here, it would make solution more robust

Reply:
@ishakhsuvarov I've remade this check, please take a look if I understand you correctly

Reply:
Ideally, I'd compare with `0`. `bccomp` would just remove all the `float` precision issues.

I guess this is out of scope for the current solution